### ISSUE TITLE
[Progress Report] Implementation of final decision layer

### ISSUE THREAD

**Issue by rightnknow** [capstone496/SpeechSentiments#14]
This is to document the implement process of constructing final decision layer using both Spectrogram+CNN and audio+LSTM models. 
Current two possible flow can be:

1. CNN and LSTM model train with same input data(in different form) at the same time, tune the dense layer of both output
2. fully train two model, froze it, then do training based on dense layer of both output

Current Implementation will be option 1

**Comment by rightnknow** [comment]
From https://github.com/keras-team/keras/issues/7581
The post mentioned that only weights of the model can be transferred between Tensorflow and Keras, therefore it's still required to build the CNN or LSTM model in Keras or Tensorflow. For now I'll continue with Keras.
The New CNN model in Keras will try to have same structure as before. which is Inception V3 +2 Dense layer

**Comment by rightnknow** [comment]
Keras provide Inception V3 weight and model in their library, along with other network models such as VGG and ResNet.

Currently there's issue found in Keras inception V3 model.
http://blog.datumbox.com/the-batch-normalization-layer-of-keras-is-broken/
https://github.com/keras-team/keras/issues/9214

Current structure is the following
![qq 20190214022743](https://user-images.githubusercontent.com/19702294/52770095-2d521c80-3000-11e9-86fb-eb7b414a2b4f.png)
The structure is identical to Katherine's model in Tensorflow.
However due to implementation method. The Keras inception v3 takes in image with # of channel = 3
Here, we do modification to the input     
```
img_input = Input(shape=(image_size, image_size, 1))
    img_conc = Concatenate()([img_input, img_input, img_input])
    base_model = InceptionV3(weights='imagenet', include_top=True,input_tensor = img_conc)
```
By this we cast a grayscale image to rgb channel and feed it into the image.(I also tried with reading the grayscale image as rgb, it doesn't change much in terms of result)

However, I face a strange bug where the network itself doesn't seem to learn from the inception v3 network.

```
    img_input = Input(shape=(image_size, image_size, 1))
    img_conc = Concatenate()([img_input, img_input, img_input])
    base_model = InceptionV3(weights='imagenet', include_top=True,input_tensor = img_conc)
    #for layer in base_model.layers:
    #    layer.trainable = False
    firstLayer = base_model.output
    secondLayer = Dropout(0.5)(firstLayer)
    outputLayer = Dense(6,activation='softmax')(secondLayer)
    model = Model(input=base_model.input,output = outputLayer )

    model.compile(loss='categorical_crossentropy',
                  optimizer='rmsprop',
                  metrics=['accuracy'])
    plot_model(model, to_file=r'C:\Users\zhanglichuan\Desktop\ECE496\lstm\model.png', show_shapes=True)
    history = model.fit(X_train_image, Y_train, epochs=30, validation_split=0.25)
```

![qq 20190214024211](https://user-images.githubusercontent.com/19702294/52770829-29bf9500-3002-11e9-9b78-9e0a5838fa79.png)
![qq 20190214024024](https://user-images.githubusercontent.com/19702294/52770831-2c21ef00-3002-11e9-8cd8-13cd14c03dd1.png)

From graph we see that although training loss is decreasing, validation loss is not. Also we get val accuracy around 20% which is very close to random guessing.

I've tried with various method like changing the size of input, change number of layers. However results are similar or even worse. 

Here's result after adding another dense layer of size 200:
![qq 20190214030441](https://user-images.githubusercontent.com/19702294/52771965-52955980-3005-11e9-8f38-3e6f956daaf3.png)
![qq 20190214030524](https://user-images.githubusercontent.com/19702294/52771998-680a8380-3005-11e9-8e13-ca63c62e8d84.png)
The result is worse compare to previous model.

Inception V3 is not the only model that can classify image. I also tried with VGG16 model. Similar result achieved.

Possible error:
1. Resizing of input spectrogram
2. Data Augmentation
3. Well, bugs in keras application

For the potential bug exist in Batch Normalization Layer of Inception v3 model,
I apply a custom patch from 
https://github.com/datumbox/keras/tree/fork/keras2.2.4
Detail explanation can be find here:
http://blog.datumbox.com/the-batch-normalization-layer-of-keras-is-broken/

Here's result
![qq 20190214033851](https://user-images.githubusercontent.com/19702294/52773936-2203ee80-300a-11e9-90b4-bf0445366833.png)
![qq 20190214033858](https://user-images.githubusercontent.com/19702294/52773947-25977580-300a-11e9-892a-d36604f3f042.png)

**Comment by rightnknow** [comment]
After some debugging, current graph looks better now, here is the graph.
![QQ截图20190312074335](https://user-images.githubusercontent.com/19702294/54198157-2a7e0680-449c-11e9-94fc-b61ea67936ae.png)
![QQ截图20190312074348](https://user-images.githubusercontent.com/19702294/54198158-2b169d00-449c-11e9-896f-bd349d5db368.png)

231/231 [==============================] - 13s 55ms/step
test accuracy is 0.4363203467073895

**Comment by rightnknow** [comment]
Summary of the Overall Implemtation

The Implementation of the network is complete, Here's the graph of the overall structure
![QQ截图20190314062032](https://user-images.githubusercontent.com/19702294/54349302-58d91e80-4621-11e9-9042-5e47170223a3.png)

**Observation during implementation**
The combined Model have relatively high accuracy compare to individual model such as CNN or LSTM. When LSTM has test accuacy around 50 and CNN has test accuracy around 45, the combined model usually achieves about 55-60 percent accuracy.

"Key Observation"
0. Combined model does do better as expected
1. lack of data
2. Parameters not tuned well

**Problems during the implementation**
The result from the combine model is unstable, especially when I split my training set into training and validation set. This cause a decrease in accuracy. 
Random Initializaiton of the model affects learning and accuracy of the model.(getting different test accuracy each time I train the model)

Here's the loss and accuracy graph
![QQ截图20190314061917](https://user-images.githubusercontent.com/19702294/54349327-68586780-4621-11e9-949f-e76cf9d07129.png)

![QQ截图20190313071231](https://user-images.githubusercontent.com/19702294/54274857-983f3680-455f-11e9-9221-e4da12c94bef.png)

Apparently there‘s something wrong with the valiation set. I don't have an explanation for it currently.

On test set
231/231 [==============================] - 0s 2ms/step
test accuracy is 0.5541125542415685

Here's a picture of the test accuracy reaching 60%:
![IMG_3434](https://user-images.githubusercontent.com/19702294/54275002-ebb18480-455f-11e9-955a-6dd08163978f.JPG)

Project Constraint meet! However I don't suggest since we don't have enough data.
The result accuracy fluctuate between 55 to 60%.
